vividus-plugin-rest-api/src/main/java/org/vividus/bdd/steps/api/JsonResponseValidationSteps.java

```diff
@@ -224,21 +225,33 @@ public void saveElementsNumberByJsonPath(String jsonPath, Set<VariableScope> sco
     @When("I wait for presence of the element by JSON path '$jsonPath' in HTTP GET response from '$resourceUrl'"
             + " for '$duration' duration")
     public void waitForJsonFieldAppearance(String jsonPath, String resourceUrl, Duration duration) throws IOException
+    {
+        waitForJsonPath(jsonPath, resourceUrl, duration);
+        doesJsonPathElementsMatchRule(jsonPath, ComparisonRule.GREATER_THAN, 0);
+    }
+
+    private void waitForJsonPath(String jsonPath, String resourceUrl, Duration duration) throws IOException
     {
         long durationInMillis = duration.toMillis();
         long expectedTime = System.currentTimeMillis() + durationInMillis;
         long pollingTimeoutMillis = durationInMillis / DURATION_DIVIDER;
-        do
+        try
         {
-            apiSteps.whenIDoHttpRequest(HttpMethod.GET, resourceUrl);
-            if (getElementsNumber(httpTestContext.getResponse().getResponseBodyAsString(), jsonPath) > 0)
+            do
             {
-                break;
+                httpRequestExecutor.executeHttpCallSafely(HttpMethod.GET, resourceUrl, null, false);
+                if (getElementsNumber(httpTestContext.getResponse().getResponseBodyAsString(), jsonPath) > 0)
+                {
+                    break;
+                }
+                Sleeper.sleep(pollingTimeoutMillis, TimeUnit.MILLISECONDS);
             }
-            Sleeper.sleep(pollingTimeoutMillis, TimeUnit.MILLISECONDS);
+            while (System.currentTimeMillis() <= expectedTime);
+        }
+        finally
+        {
+            httpTestContext.releaseRequestData();
         }
```

Comment:
this logic is duplicated: here and `HttpRequestExecutor`

Reply:
How this could be avoided? separate method with Consumers as input?

Reply:
yes, lambdas will save the world